### ISSUE TITLE
No logging for the output of VM session command

### ISSUE THREAD

**Issue by YongxueHong** [avocado-framework/avocado-vt#2822]
There are no logfiles to recording the output of VM session, like the prefix session-avocado-vt- filename which will log the output of the guest commands to help debug automation bug or other checking and if restore the changed codes to previous, it works on logging the output of session commands.

Refer: https://github.com/avocado-framework/avocado-vt/pull/2727#issuecomment-699732188

**Comment by pevogam** [comment]
Hi @YongxueHong, do you mind being a bit more specific what you mean here? For instance, the `aexpect.remote` functions have a logging function and filename that could be specified:
```python
def login_from_session(session, log_filename=None, log_function=None,
                       timeout=240, internal_timeout=10, interface=None):
```

**Comment by YongxueHong** [comment]
Hi @pevogam 
I mean line 311~312 of  https://github.com/avocado-framework/avocado-vt/pull/2727/files#diff-c43bafa8b2d370504f3281b4659f562a5a9979d05d7a1f8865a53c06a8fdca47, I didn't find the session log file under the path `../job-results/job-xxxx-xx-xxxx.xx-xxxxxx/` after running cases.
For the original codes, there will be generated the session log files under the related directory of testing result.
Thanks.

**Comment by pevogam** [comment]
The current way this is handled in aexpect is like so:
```python
    if log_filename:
        session.set_output_func(log_function)
        session.set_output_params((log_filename,))
        session.set_log_file(os.path.basename(log_filename))
```
In other words, we generalize to any filename and log function there and you should be able to use something as simple as
```python
remote_login(..., log_filename=utils_logfile.get_log_filename(log_filename), log_function=utils_logfile.log_line, timeout=...)
```
on the VT side. So if I understand this right, we need a fix like this one on the calling side.

**Comment by YongxueHong** [comment]
Hi, @pevogam 
For this session logging function is important to our team since it is a very useful way to check the output of the session then debug the automation issue.
From https://github.com/avocado-framework/avocado-vt/pull/2095 patch, I saw that we migrated `remote_login` to `aexpect.remote.remote_login`, but the latter changes the behavious about session logging function:

virttest.remote:
```
def remote_login(client, host, port, username, password, prompt, linesep="\n",
                 log_filename=None, timeout=10, interface=None, identity_file=None,
                 status_test_command="echo $?", verbose=False, bind_ip=None,
                 preferred_authenticaton='password',
                 **user_known_hosts_file='/dev/null'):
...
    if log_filename:
        log_file = utils_logfile.get_log_filename(log_filename)
        session.set_output_func(utils_logfile.log_line)
        session.set_output_params((log_file,))
        session.set_log_file(os.path.basename(log_file))
```
aexpect.remote:
```
def remote_login(client, host, port, username, password, prompt, linesep="\n",
                 log_filename=None, log_function=None, timeout=10,
                 interface=None, identity_file=None,
                 status_test_command="echo $?", verbose=False, bind_ip=None,
                 preferred_authenticaton='password',
                 user_known_hosts_file='/dev/null'):
...
    if log_filename:
        session.set_output_func(log_function)
        session.set_output_params((log_filename,))
        session.set_log_file(os.path.basename(log_filename))
```
The caller:
virttest.virt_vm:
```
    def login(self, nic_index=0, timeout=LOGIN_TIMEOUT,
              username=None, password=None):
        ...
        log_filename = ("session-%s-%s-%s.log" %
                        (self.name, time.strftime("%m-%d-%H-%M-%S"),
                         utils_misc.generate_random_string(4)))
        session = remote.remote_login(client, address, port, username,
                                      password, prompt, linesep,
                                      log_filename, timeout,
                                      interface=neigh_attach_if)
```

so this patch https://github.com/avocado-framework/avocado-vt/pull/2095 introduces a regression bug about the session logging function.

Could you help me to solve it?
Thanks.
cc @ldoktor

**Comment by ldoktor** [comment]
Hello @YongxueHong, ideally the log function would be implemented in aexpect. @pevogam what do you think about that? But for now we can simply change the default None (see #2984) to our usual value in order to bring back our old behaviour, what do you think?

**Comment by pevogam** [comment]
@YongxueHong So if I understand correctly this is actually reporting a regression in `virttest.virt_vm` due to the positional arguments, is this correct? If so can't we just propose a PR like
```diff
     def login(self, nic_index=0, timeout=LOGIN_TIMEOUT,
               username=None, password=None):
         ...
         log_filename = ("session-%s-%s-%s.log" %
                         (self.name, time.strftime("%m-%d-%H-%M-%S"),
                          utils_misc.generate_random_string(4)))
         session = remote.remote_login(client, address, port, username,
                                       password, prompt, linesep,
-                                      log_filename, timeout,
+                                      log_filename, timeout=timeout,
                                       interface=neigh_attach_if)
```
?

> Hello @YongxueHong, ideally the log function would be implemented in aexpect. @pevogam what do you think about that?

I am not sure how generic logging utilities like the ones from `utils_logfile` relate to aexpect in particular. The current default value `None` is much cleaner and gives everybody the freedom to provide whatever logging and logfiles they want.

> But for now we can simply change the default None (see #2984) to our usual value in order to bring back our old behaviour, what do you think?

This type of change will pull us back a step since what we want is to eventually rid ourselves of the old `remote.py` file there completely and fix any detected problems like this one. Amending it the way it is done in this PR is going to hide other potential such problems which are already easy to fix (e.g. like above) and that are better be done with.

**Comment by ldoktor** [comment]
Hello @pevogam, the difference is that before Avocado-vt only supported `log_filename` and it created the `log_function` when needed. This was very convenient but you are right that less flexible than having an access to a custom `log_function`.

Your PR made the `log_function` available, but the problem is it's only used when `log_filename` is set (and uses it as an argument) although generic `log_function` might not require this. Second problem is that when a user simply wants to write to a file, they need to specify both, the `log_filename` and `log_function`, otherwise it's not doing anything (which is confusing as supporting `log_filename` one would assume the file will be created).

To make things explicit, it'd be nice to either remove the `log_function` (in aexpect) and only let users to write to a file or to remove the `log_filename` and let people to set a generic `log_function` (probably a lambda with default parameters). Those two solutions would be IMO easier to understand, first one being super easy to use, the other even more flexible than the current solution while requiring more work from users (which might indicate I'm more inclined to the simpler solution, unless someone is using custom functions).

**Comment by ldoktor** [comment]
@pevogam and yes, my PR is a step backward, more like an example on how to deal with this situation. Real fix should go to aexpect and avocado-vt should adjust (after a transition period)

**Comment by pevogam** [comment]
> Hello @pevogam, the difference is that before Avocado-vt only supported `log_filename` and it created the `log_function` when needed. This was very convenient but you are right that less flexible than having an access to a custom `log_function`.
> 
> Your PR made the `log_function` available, but the problem is it's only used when `log_filename` is set (and uses it as an argument) although generic `log_function` might not require this. Second problem is that when a user simply wants to write to a file, they need to specify both, the `log_filename` and `log_function`, otherwise it's not doing anything (which is confusing as supporting `log_filename` one would assume the file will be created).

Indeed, so it was a half-baked change then which only made sure to preserve the previous logic while offering a rather superficial freedom of choosing the `log_function`. Good points all around. So ultimately we have to refactor this snippet:
```
    if log_filename:
        session.set_output_func(log_function)
        session.set_output_params((log_filename,))
        session.set_log_file(os.path.basename(log_filename))
```
to
```
    if log_function:
        session.set_output_func(log_function)
    if log_filename:
        session.set_output_params((log_filename,))
        session.set_log_file(os.path.basename(log_filename))
```
It doesn't seem like much work to do or am I missing something? I must admit I am proposing this purely from a superficial point of view without taking deeper look into the `session.set_*` methods.

> To make things explicit, it'd be nice to either remove the `log_function` (in aexpect) and only let users to write to a file or to remove the `log_filename` and let people to set a generic `log_function` (probably a lambda with default parameters). Those two solutions would be IMO easier to understand, first one being super easy to use, the other even more flexible than the current solution while requiring more work from users (which might indicate I'm more inclined to the simpler solution, unless someone is using custom functions).

Indeed we want the best from both worlds - freedom to choose and automation for simple default choices. Why is offering both arguments for optional setting like the proposal above not a viable option?

> @pevogam and yes, my PR is a step backward, more like an example on how to deal with this situation. Real fix should go to aexpect and avocado-vt should adjust (after a transition period)

Isn't the [diff above](https://github.com/avocado-framework/avocado-vt/issues/2822#issuecomment-811015421) an easy way to avoid taking this step backward at least for the current PR and issue?

**Comment by pevogam** [comment]
> Second problem is that when a user simply wants to write to a file, they need to specify both, the `log_filename` and `log_function`, otherwise it's not doing anything (which is confusing as supporting `log_filename` one would assume the file will be created).

I see, I guess then it is also confusion arising from the way the `session.set_*` API is structured. Just from looking at `set_output_func`, `set_output_params`, and `set_log_file` as public methods I would assume they are independent of each other or would at least call each other internally if otherwise. If I am wrong then the problem runs deeper indeed.

**Comment by YongxueHong** [comment]
FYI.
CC @yanan-fu

**Comment by ldoktor** [comment]
Yes, @pevogam, the problem is that aexpect was strictly specialized tool with no magic (almost). The `set_output_func` and `set_output_params` functions were IMO a bit redundant as the caller could just create an ad hock lambda function and store it as a function.

Anyway we are not trying to simplify the aexpect API. Given that so much code was added to aexpect, maybe it'd make sense to a sample write-to-file function into it as well and use it when no function is specified but output file is. That would be the most flexible solution while very convenient for the simple write-to-file usecase.

**Comment by pevogam** [comment]
> Yes, @pevogam, the problem is that aexpect was strictly specialized tool with no magic (almost). The `set_output_func` and `set_output_params` functions were IMO a bit redundant as the caller could just create an ad hock lambda function and store it as a function.

I totally agree here and very often I have seen people add to Avocado VT their personalized write-to-file functions all the time (where the latter is a metaphor for all sorts of private solutions of own problems that don't generalize to the VT plugin). I guess the idea here was simply to migrate the functionality and generalize a bit as a bonus wherever possible like in this case.

> Anyway we are not trying to simplify the aexpect API. Given that so much code was added to aexpect, maybe it'd make sense to a sample write-to-file function into it as well and use it when no function is specified but output file is. That would be the most flexible solution while very convenient for the simple write-to-file usecase.

The code that was added to aexpect is still not really a lot of magic, I think it is still reasonable to extend the session classes the way we did and to add some file copy support on top of sessions the way we did.

Your proposed solution is probably the best to minimize more potentials changes. Should we move this to a PR? @YongxueHong, are you interested in proposing the PR suggested by @ldoktor? As a very temporary solutions I still think that many of the ones above remain legitimate.

**Comment by YongxueHong** [comment]
Hi @ldoktor 
Maybe your patch https://github.com/avocado-framework/avocado-vt/pull/2984 does not work, since https://github.com/avocado-framework/avocado-vt/pull/2984#issuecomment-828122188 
CC @yanan-fu
